authentikos/test/integ-simple.sh

```diff
@@ -32,42 +32,49 @@ get_tokeninfo() {
   curl -sSfL "https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=$token"
 }
 
-run_test() {
+run_test() { (
   set -x
 
   local tokeninfo="$1"
 
+  echo "Test 'tokeninfo' response body exists"
+  test -n "$tokeninfo"
+
   echo "Test 'error' is null"
-  echo "$tokeninfo" | jq -r '.error' | xargs -r -I% test % = "null"
+  test "$(echo "$tokeninfo" | jq -r '.error')" = "null"
 
   echo "Test 'issued_to' exists"
-  echo "$tokeninfo" | jq -r '.issued_to' | xargs -r test -n
+  test -n "$(echo "$tokeninfo" | jq -r '.issued_to')"
 
   echo "Test 'audience' exists"
-  echo "$tokeninfo" | jq -r '.audience' | xargs -r test -n
+  test -n "$(echo "$tokeninfo" | jq -r '.audience')"
 
   echo "Test 'user_id' exists"
-  echo "$tokeninfo" | jq -r '.user_id' | xargs -r test -n
+  test -n "$(echo "$tokeninfo" | jq -r '.user_id')"
 
   echo "Test 'email' exists"
-  echo "$tokeninfo" | jq -r '.email' | xargs -r test -n
+  test -n "$(echo "$tokeninfo" | jq -r '.email')"
 
   echo "Test 'expires_in' greater than 0"
-  echo "$tokeninfo" | jq -r '.expires_in' | xargs -r -I% test % -lt 0
+  test "$(echo "$tokeninfo" | jq -r '.expires_in')" -gt 0
 
   echo "Test 'scope' includes 'userinfo.email', 'cloud-platform', and 'openid'"
   echo "$tokeninfo" | jq -r '.scope' |
-    grep -w "openi" |
+    grep -w "openid" |
     grep -w "https://www.googleapis.com/auth/cloud-platform" |
     grep -w "https://www.googleapis.com/auth/userinfo.email" >/dev/null
-}
+); }
 
 main() {
   kubectl create secret generic service-account --from-file="service-account.json=$GOOGLE_APPLICATION_CREDENTIALS"
   kubectl apply --filename="$ROOT/testdata/authentikos-simple.yaml"
   kubectl wait deployment authentikos --for="condition=available" --timeout="$timout"
 
   run_test "$(with_timeout get_tokeninfo "$timout")"
+  exit_code="$?"
```

Comment:
We're not reaching this line because the previous line fails and exit on error is enabled.  I spent some time messing around with the behavior of `set -e` and it is pretty confusing when trying to handle errors that occur in nested functions so I think my previous advice was bad.
Now I think `trap`ping the log dump is the easiest way to ensure we do so after tests complete or fail.

Reply:
I ended up wrapping the function body in a subshell. This will allow the function to exit but not the parent shell [[ref](https://stackoverflow.com/a/58964551/4279440)].

https://github.com/istio/test-infra/blob/a21aba25a5e916fc738673a60713c73362b5e962/authentikos/test/integ-simple.sh#L35

Reply:
Hmm that does seem like it should work, but I still don't see any logs from Authentikos. Perhaps Authentikos has no logs? (I'm trying to debug locally, but I'm having problems getting a kind cluster up on my local machine.)

Reply:
The logging is not the best but it exists. When I run locally using a personal SA I get the following logs.

```
2021/05/26 22:01:31 expired: false; token expiry (minus delta): 2021-05-26 22:56:30.520797031 +0000 UTC m=+3309.168705502; next reconcile: 2021-05-26 22:31:31.520799684 +0000 UTC m=+1810.168708140
2021/05/26 22:01:31 creating secret: authentikos-token in namespace: default
2021/05/26 22:01:31 2021-05-26T22:01:31Z: reconcile complete; secrets: 1; errors: 0; next reconcile: 30m 
```

Reply:
I started my own test PR to mess around with this in CI. I'm not seeing logs from Authentikos, but it also seems like the subshell is not preventing an early exit as expected: https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3368/integ-test-authentikos_test-infra/1400242374365417472
(Note that there is no echoed "Dumping logs before exit".)

Reply:
Thanks for debugging. Let me verify this locally, but I think I just need to unset the exit on non-zero status code option (i.e. `set +e`)  in the parent shell to prevent the premature exit.

https://github.com/istio/test-infra/blob/ce16c4297dc86fe77e85ac9b3ce7fa02b1b50018/tools/automator/automator.sh#L331-L337

Reply:
@cjwagner – I made the modifications mentioned above and verified locally that the Authentikos logs are emitted.

Reply:
Cool, now we're getting some logs:
> 2021/06/03 02:38:58 2021-06-03T02:38:58Z: reconcile complete; secrets: 0; errors: 1; next reconcile: 30m
2021/06/03 02:38:58 errors: default:maximum tries: 5 exceeded to force refresh token

Unfortunately it doesn't look like we're logging the relevant error messages. I made a PR to add some more logging: https://github.com/istio/test-infra/pull/3370